db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/FromBaseTable.java

```diff
@@ -3871,8 +4014,33 @@ private PredicateList translateBetweenHelper(PredicateList predList) throws Stan
             }
         }
         if (translated) {
-            newList.classify(this, getTrulyTheBestAccessPath().getConglomerateDescriptor(), true);
+            newList.classify(this, getTrulyTheBestAccessPath(), true);
         }
         return newList;
     }
+
+    @Override
+    public boolean hasIndexPrefixIterator() { return numUnusedLeadingIndexFields > 0; }
+
+    @Override
+    public boolean indexPrefixIteratorAllowed(Optimizer optimizer) {
+        if (disableIndexPrefixIteration)
+            return false;
+        LanguageConnectionContext lcc = getLanguageConnectionContext();
+        if (lcc.alwaysAllowIndexPrefixIteration())
+            return true;
+
+        if (getCompilerContext().getDisablePrefixIteratorMode())
+            return false;
+
+        // We must have statistics collected on the base table to pick IndexPrefixIteratorMode
+        // because the operation may get expensive if the number of values is underestimated.
+        // Also, building of the MultiRowRangeFilter to handle the iteration may get expensive
+        // for large numbers of values, so limit it to 20000 on control and 1000000 on Spark
+        // for now.
+        long maxPrefixIteratorValues = optimizer.isForSpark() ? 1000000 : 20000;
```

Comment:
Slightly off topic, but where do these magic numbers come from? does it make sense to have them correlating with some environment parameters? e.g. for Spark we take the number of cores into account `c` and available memory `m` to come up with something like: `1000000*m*c`

Reply:
This is a very good point.  The more parallel Spark tasks we have, then more the work will be spread out to different tasks.  For 8 tasks, each task may only process 1/8 of the values.
I have added:
`    // The maximum number of estimated first index column values
    // per parallel task to support for IndexPrefixIteratorMode.
    private final long MAX_PER_PARALLEL_TASK_INDEX_PREFIX_VALUES = 200000;

    // The absolute maximum number of index values for IndexPrefixIteratorMode.
    // Iterating through these values will be divided by n tasks.
    private final long MAX_ABSOLUTE_INDEX_PREFIX_VALUES = 50000000;`
and
`        long sparkMaxPrefixIteratorValues = Long.min(MAX_ABSOLUTE_INDEX_PREFIX_VALUES,
                                                     parallelism * MAX_PER_PARALLEL_TASK_INDEX_PREFIX_VALUES);
        long maxPrefixIteratorValues = optimizer.isForSpark() ? sparkMaxPrefixIteratorValues :
                                       sparkRowThreshold;`

The value 20000 comes from calling getDetermineSparkRowThreshold().  I added this in instead of using a constant since this value is configurable:
`long sparkRowThreshold = lcc.getOptimizerFactory().getDetermineSparkRowThreshold();`

For the value 1000000, this is just a conservative value since I don't want to use too high of a value and make the optimization too aggressive in the first pass.  I've adjusted this to 200000 per parallel task, but maybe we should make it 150000.  I will try some tests.